src/wayland/xdg_activation/mod.rs

```diff
@@ -135,6 +123,8 @@ pub struct XdgActivationTokenData {
     /// You can use this do ignore tokens based on time.
     /// For example you coould ignore all tokens older that 5s.
     pub timestamp: Instant,
+    /// Additional user data attached
+    pub user_data: Arc<UserDataMap>,
```

Comment:
I mean it's totally optional. I can also drop the commit given, that the token itself implements `Hash`. Downstream can just implement their own `HashMap` to store additional data. But that seems unnecessary given we already have such a map.

I don't think there is any harm in allowing this to exist as well.

I guess an alternative would be an associated type on the `XdgActivationHandler` and make `token_created` return a `Option<Self::UserData>`.

Reply:
>I mean it's totally optional. I can also drop the commit given, that the token itself implements Hash. Downstream can just implement their own HashMap to store additional data. But that seems unnecessary given we already have such a map.

I think downstream should just have a state for clients they manage, since activation token is generally associated with the client created it. It's just downstream already has `&mut State` and it probably should have `ClientId` based handling for some stuff already (e.g. keyboard layout switches, etc).

In general, if you can't find a good use case for something just don't add it, since it blows the API and makes it harder to remove things.

Reply:
> I think downstream should just have a state for clients they manage, since activation token is generally associated with the client created it. It's just downstream already has &mut State and it probably should have ClientId based handling for some stuff already (e.g. keyboard layout switches, etc).

There is other data you might want to associate with a token, that is not related to the client. E.g. which workspace was active, when the token was created, to spawn new clients on that workspace later. Given a single client can have multiple surfaces on different workspaces, the `ClientId` is no of any use here.

> In general, if you can't find a good use case for something just don't add it, since it blows the API and makes it harder to remove things.

There you have a use-case. I personally don't like having another `HashMap` for the same purpose in my compositor. I would like the xdg-activation implementation to manage data associated with the token for me (as it already has to do anyway).

`UserDataMap` is already prevalent through the code-base anyway as it is a handy way to associate arbitrary data. If we want to phase that out in general we should have an issue about that and a lengthy discussion. Possibly we can at least write down some rules, where we see it's use appropriate and where not. Though in this case the problem could also be solved by generics, so I'll opt for that instead.

> makes it harder to remove things.

I don't think smithay is at the point where we care about API stability to this extent, given I just suggested discussing about `UserDataMap` in general.

Reply:
>There is other data you might want to associate with a token, that is not related to the client. E.g. which workspace was active, when the token was created, to spawn new clients on that workspace later. Given a single client can have multiple surfaces on different workspaces, the ClientId is not of any use here.

Hm, on the client side we already use data, but mostly RO (until someone tries to abuse it). I guess it's fine the way it is for now given that `UserDataMap` is effectively RO...

In general when activating with workspaces you care about `currently focused` and where the new window is being activated, you don't need the data on which work space it was created. But workspaces are hard, because they need help client side as well. There was
some discussion somewhere on that matter.